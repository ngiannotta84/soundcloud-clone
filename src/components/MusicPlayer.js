/* eslint-disable jsx-a11y/media-has-caption */
import React, { useState, useRef, useEffect } from "react";
import PropTypes from "prop-types";
import "../styles/musicplayer.css";

const MusicPlayer = ({ playlist, playlistIndex, setPlaylistIndex }) => {
  if (!playlist || playlist.length === 0) return null;
  const [playing, setPlaying] = useState(false);
  const [time, setTime] = useState(0);
  const [duration, setDuration] = useState(0);
  const [volume, setVolume] = useState(0.5);
  const audioRef = useRef(null);

  const song = playlist[playlistIndex];

  const handlePlaying = (isPlaying) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev);
    } else {
      setPlaying(isPlaying);
    }
  };

  const skipSong = (forwards = true) => {
    if (forwards) {
      if (playlistIndex === playlist.length) {
        setPlaylistIndex(0);
      } else {
        setPlaylistIndex((prev) => prev + 1);
      }
    } else if (playlistIndex === 0) {
      setPlaylistIndex(playlist.length);
    } else {
      setPlaylistIndex((prev) => prev - 1);
    }
  };

  const calcTime = (seconds) => {
    let sec = seconds;
    let min = 0;
    while (sec >= 60) {
      min += 1;
      sec -= 60;
    }
    if (sec < 10) {
      sec = `0${sec}`;
    }
    return `${min}:${sec}`;
  };

  const handleLoadedMetaData = () => {
    if (audioRef.current) {
      const seconds = Math.round(audioRef.current.duration);
      setDuration(seconds);
      if (playing) {
        audioRef.current.play();
      }
    }
  };

  const handleProgressBarChange = (e) => {
    const newTime = Number(e.target.value);
    if (audioRef.current) {
      audioRef.current.currentTime = newTime;
    }
    setTime(newTime);
  };

  const handleVolume = (e) => {
    const value = Number(e.target.value / 100);
    if (value > 1) {
      setVolume(1);
    } else if (value < 0) {
      setVolume(0);
    } else {
      setVolume(value);
    }
  };

  useEffect(() => {
    if (audioRef.current) {
      if (playing) {
        audioRef.current.play();
      } else {
        audioRef.current.pause();
      }
    }
  }, [playing, audioRef, playlistIndex]);

  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = volume;
    }
  }, [volume, audioRef]);

  useEffect(() => {
    const updateTimer = setInterval(() => {
      if (audioRef.current) {
        const seconds = Math.round(audioRef.current.currentTime);
        setTime(seconds);
      }
    }, 500);
    return () => clearInterval(updateTimer);
  }, [time]);

  return (
    <>
      <audio
        src={song.audio}
        ref={audioRef}
        onLoadedMetadata={handleLoadedMetaData}
      />
      <div className="music-player">
        <div className="music-player__left">
          <img
            src={song.image}
            alt={`${song.albumName} cover art`}
            className="music-player__cover-art"
          />
          <div className="music-player__info">
            <h2>{song.artistName}</h2>
            <h3>{song.songName}</h3>
          </div>
        </div>
        <div className="music-player__center">
          <input
            type="range"
            name="time"
            min={0}
            max={duration}
            value={time}
            onChange={handleProgressBarChange}
            className="music-player__progress-bar"
          />
          <div className="music-player__center__under">
            <p>{calcTime(time)}</p>
            <div>
              <button type="button" onClick={() => skipSong(false)}>
                Back
              </button>
              <button type="button" onClick={() => handlePlaying()}>
                Play/Pause
              </button>
              <button type="button" onClick={() => skipSong()}>
                Forwards
              </button>
            </div>
            <p>{calcTime(duration)}</p>
          </div>
        </div>
        <input
          type="range"
          onChange={handleVolume}
          value={volume * 100}
          min={0}
          max={100}
          name="volume"
        />
      </div>
    </>
  );
};

MusicPlayer.defaultProps = {
  playlist: null,
};

MusicPlayer.propTypes = {
  playlist: PropTypes.arrayOf(
    PropTypes.shape({
      image: PropTypes.string,
      artistName: PropTypes.string,
      albumName: PropTypes.string,
      audio: PropTypes.string,
      songName: PropTypes.string,
    })
  ),
  playlistIndex: PropTypes.number.isRequired,
  setPlaylistIndex: PropTypes.func.isRequired,
};

export default MusicPlayer;
