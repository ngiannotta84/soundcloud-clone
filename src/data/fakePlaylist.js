import card from "./images/card.png";
import dodl from "./images/Final Render_Cropped 3000.png";
import slumber from "./songs/Slumber.flac";
import careless from "./songs/Your Careless Embrace.flac";
import restless from "./songs/Restless Thoughts.flac";
import detunedLove from "./songs/Detuned Love.flac";
import falling from "./songs/Falling into the Void.flac";

const playlist = [
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: slumber,
    songName: "Slumber",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: careless,
    songName: "Your Careless Embrace",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: restless,
    songName: "Restless Thoughts",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: detunedLove,
    songName: "Detuned Love",
  },
  {
    image: dodl,
    artistName: "Jenico",
    albumName: "Dreaming of Detuned Love",
    audio: falling,
    songName: "Falling into the Void",
  },
];

export default playlist;
