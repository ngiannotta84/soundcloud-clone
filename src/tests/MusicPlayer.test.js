import React from "react";
import { render, screen } from "@testing-library/react";
import MusicPlayer from "../components/MusicPlayer";
import fakePlaylist from "./testData/fakePlaylist";

describe("MusicPlayer", () => {
  const validProps = {
    playlist: fakePlaylist,
    playlistIndex: 1,
    setPlaylistIndex: jest.fn(),
    removeFromPlaylist: jest.fn(),
  };
  const audioPlay = jest.fn();
  const audioPause = jest.fn();

  jest
    .spyOn(window.HTMLMediaElement.prototype, "play")
    .mockImplementation(audioPlay);
  jest
    .spyOn(window.HTMLMediaElement.prototype, "pause")
    .mockImplementation(audioPause);

  test("snapshot", () => {
    const { asFragment } = render(<MusicPlayer {...validProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(<MusicPlayer {...validProps} />);
    });

    test("renders correctly", () => {
      const song = validProps.playlist[validProps.playlistIndex];
      const coverArt = screen.getByAltText(/cover art/i);

      expect(screen.getByTestId("audio")).toHaveAttribute("src", song.audio);
      expect(coverArt).toHaveAttribute("alt", `${song.albumName} cover art`);
      expect(coverArt).toHaveAttribute("src", song.image);
      expect(
        screen.getByText(`${song.artistName} - ${song.songName}`)
      ).toBeInstanceOf(HTMLHeadingElement);
      expect(screen.getByTestId("progressBar")).toHaveAttribute(
        "type",
        "range"
      );
      expect(screen.getByAltText(/skip backwards/i)).toBeInTheDocument();
      expect(screen.getByAltText(/play/i)).toBeInTheDocument();
      expect(screen.getByAltText(/skip forwards/i)).toBeInTheDocument();
      expect(screen.getByAltText(/volume/i)).toBeInTheDocument();
      expect(screen.getByAltText(/playlist/i)).toBeInTheDocument();
    });
  });
});
