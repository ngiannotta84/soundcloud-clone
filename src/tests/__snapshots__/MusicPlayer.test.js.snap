// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MusicPlayer snapshot 1`] = `
<DocumentFragment>
  <audio
    data-testid="audio"
    src="audio 2"
  />
  <div
    class="music-player"
  >
    <div
      class="music-player__left"
    >
      <img
        alt="album 2 cover art"
        class="music-player__cover-art"
        src="image 2"
      />
      <div
        class="music-player__info"
      >
        <h2>
          artist 2
        </h2>
        <h3>
          song 2
        </h3>
      </div>
    </div>
    <div
      class="music-player__center"
    >
      <input
        class="music-player__progress-bar"
        data-testid="progressBar"
        max="0"
        min="0"
        name="time"
        type="range"
        value="0"
      />
      <div
        class="music-player__center__under"
      >
        <p>
          0:00
        </p>
        <div>
          <button
            class="music-player__skip-backwards"
            type="button"
          >
            <img
              alt="skip backwards"
              src="skip.png"
            />
          </button>
          <button
            class="music-player__play-pause"
            type="button"
          >
            <img
              alt="pause"
              src="play.png"
            />
          </button>
          <button
            class="music-player__skip-forwards"
            type="button"
          >
            <img
              alt="skip forwards"
              src="skip.png"
            />
          </button>
        </div>
        <p>
          0:00
        </p>
      </div>
    </div>
    <div
      class="music-player__right"
    >
      <div
        class="music-player__volume"
      >
        <button
          type="button"
        >
          <img
            alt="volume"
            src="volume-medium.png"
          />
        </button>
      </div>
      <div
        class="music-player__playlist"
      >
        <button
          type="button"
        >
          <img
            alt="playlist"
            src="menu.png"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
